plotter/contour.go

```diff
@@ -0,0 +1,488 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:generate ./list
+
+package plotter
+
+import (
+	"image/color"
+	"math"
+	"sort"
+
+	"github.com/gonum/plot"
+	"github.com/gonum/plot/palette"
+	"github.com/gonum/plot/vg"
+	"github.com/gonum/plot/vg/draw"
+)
+
+// Contour implements the Plotter interface, drawing
+// a contour plot of the values in the GridXYZ field. The
+// row index is from top to bottom.
+type Contour struct {
+	GridXYZ GridXYZ
+
+	// Levels describes the contour heights to plot.
+	Levels []float64
+
+	// LineStyle is the style of the contour lines.
+	draw.LineStyle
+
+	// Palette is the color palette used to render
+	// the heat map.
+	Palette palette.Palette
+
+	// Underflow and Overflow are colors used to draw
+	// contours outside the dynamic range defined
+	// by Min and Max.
+	Underflow color.Color
+	Overflow  color.Color
+
+	// Min and Max define the dynamic range of the
+	// heat map.
+	Min, Max float64
+}
+
+// NewContour creates as new contour plotter for the given data,
+// using the provided palette.
+func NewContour(g GridXYZ, levels []float64, p palette.Palette) *Contour {
+	var min, max float64
+	type minMaxer interface {
+		Min() float64
+		Max() float64
+	}
+	switch g := g.(type) {
+	case minMaxer:
+		min, max = g.Min(), g.Max()
+	default:
+		min, max = math.Inf(1), math.Inf(-1)
+		c, r := g.Dims()
+		for i := 0; i < c; i++ {
+			for j := 0; j < r; j++ {
+				v := g.Z(i, j)
+				if math.IsNaN(v) {
```

Comment:
Other plotters return an error if there are NaN values (scatter). It doesn't need to be fixed now, but we should be consistent about it.